pkg/controller/master-controller-manager/serviceaccount-projectbinding-controller/controller.go

```diff
@@ -14,89 +14,99 @@ See the License for the specific language governing permissions and
-	r := &reconcileServiceAccountProjectBinding{Client: mgr.GetClient()}
+// reconcileServiceAccountProjectBinding reconciles User objects.
+type reconcileServiceAccountProjectBinding struct {
+	ctrlruntimeclient.Client
+
+	log *zap.SugaredLogger
+}
+
+func Add(mgr manager.Manager, log *zap.SugaredLogger) error {
+	r := &reconcileServiceAccountProjectBinding{
+		Client: mgr.GetClient(),
+
+		log: log,
+	}
 	// Create a new controller
 	c, err := controller.New(controllerName, mgr, controller.Options{Reconciler: r})
 	if err != nil {
 		return err
 	}
 
 	// Watch for changes to User
-	return c.Watch(&source.Kind{Type: &kubermaticv1.User{}}, &handler.EnqueueRequestForObject{})
-}
+	isServiceAccount := predicate.NewPredicateFuncs(func(object ctrlruntimeclient.Object) bool {
+		return kubermaticv1helper.IsProjectServiceAccount(object.GetName())
+	})
 
-// reconcileServiceAccountProjectBinding reconciles User objects.
-type reconcileServiceAccountProjectBinding struct {
-	client.Client
+	if err = c.Watch(&source.Kind{Type: &kubermaticv1.User{}}, &handler.EnqueueRequestForObject{}, isServiceAccount); err != nil {
+		return err
+	}
+
+	return nil
 }
 
 func (r *reconcileServiceAccountProjectBinding) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
-	resourceName := request.Name
-	// project service account
-	if strings.HasPrefix(resourceName, ServiceAcountName) {
-		if err := r.ensureServiceAccountProjectBinding(ctx, resourceName); err != nil {
-			logger := log.Logger.With("controller", controllerName)
-			logger.Errorw("failed to reconcile in controller", "error", err)
-			return reconcile.Result{}, err
-		}
+	log := r.log.With("serviceaccount", request.Name)
+
+	err := r.ensureServiceAccountProjectBinding(ctx, log, request.Name)
+	if err != nil {
+		log.Errorw("failed to reconcile", zap.Error(err))
 	}
 
-	return reconcile.Result{}, nil
+	return reconcile.Result{}, err
 }
 
-func (r *reconcileServiceAccountProjectBinding) ensureServiceAccountProjectBinding(ctx context.Context, saName string) error {
+func (r *reconcileServiceAccountProjectBinding) ensureServiceAccountProjectBinding(ctx context.Context, log *zap.SugaredLogger, saName string) error {
+	log.Debug("Reconciling")
+
 	sa := &kubermaticv1.User{}
-	if err := r.Get(ctx, client.ObjectKey{Namespace: metav1.NamespaceAll, Name: saName}, sa); err != nil {
+	if err := r.Get(ctx, ctrlruntimeclient.ObjectKey{Name: saName}, sa); err != nil {
 		if kerrors.IsNotFound(err) {
 			return nil
 		}
 		return err
 	}
 
 	// get project name from Owner Reference or from label
-	projectName := ""
-	for _, owner := range sa.GetOwnerReferences() {
-		if owner.APIVersion == kubermaticv1.SchemeGroupVersion.String() && owner.Kind == kubermaticv1.ProjectKindName &&
-			len(owner.Name) > 0 && len(owner.UID) > 0 {
-			projectName = owner.Name
-			break
-		}
-	}
+	projectName := sa.Spec.Project
 	if len(projectName) == 0 {
+		// TODO: Is this fallback still needed?
```

Comment:
i dont think so. The API is setting the ownerRef (now after this PR the project field) and if somebody wants to use gitops they can use the new project field.

Reply:
The RBAC controller has some code to read the project from the label. Are you sure we can remove this? I would be too scared :D

Reply:
You mean here?
https://github.com/kubermatic/kubermatic/blob/2ffbd70a7531cd9056fd5a0eb619f6a977899d29/pkg/controller/master-controller-manager/rbac/sync_resource.go#L223-L240

It checks for ownerRef first as well, which should be set by the new controller. 

I mean yeah, somebody put this check here in serviceAccount controller for some reason I guess. But it could be some legacy. I dont see anywhere in code that the `projectID` label is put on serviceaccount users(or users at all), and on dev no user has it (but thats no proof ofc).

Maybe lets remove it as a separate PR so we can revert it easily if needed